service/storage_service.cc

```diff
@@ -743,12 +850,70 @@ future<> storage_service::topology_change_coordinator_fiber(raft::server& raft,
                         auto tmptr = get_token_metadata_ptr();
                         auto bootstrap_tokens = boot_strapper::get_random_bootstrap_tokens(tmptr, num_tokens, dht::check_token_endpoint::yes);
 
-                        // Write choosen tokens through raft.
+                        auto get_sharding_info = [&] (dht::token end) -> std::pair<size_t, uint8_t> {
+                            if (bootstrap_tokens.contains(end)) {
+                                return {node.rs->shard_count, node.rs->ignore_msb};
+                            } else {
+                                // FIXME: token metadata should directly return host ID for given token. See #12279
+                                auto ep = tmptr->get_endpoint(end);
+                                if (!ep) {
+                                    // get_sharding_info is only called for bootstrap tokens
+                                    // or for tokens present in token_metadata
+                                    on_internal_error(slogger, format(
+                                        "raft topology: make_new_cdc_generation_data: get_sharding_info:"
+                                        " can't find endpoint for token {}", end));
+                                }
+
+                                auto id = tmptr->get_host_id_if_known(*ep);
+                                if (!id) {
+                                    on_internal_error(slogger, format(
+                                        "raft topology: make_new_cdc_generation_data: get_sharding_info:"
+                                        " can't find host ID for endpoint {}, owner of token {}", *ep, end));
+                                }
+
+                                auto ptr = node.topology->find(raft::server_id{id->uuid()});
+                                if (!ptr) {
+                                    on_internal_error(slogger, format(
+                                        "raft topology: make_new_cdc_generation_data: get_sharding_info:"
+                                        " couldn't find node {} in topology, owner of token {}", *id, end));
+                                }
+
+                                auto& rs = ptr->second;
+                                return {rs.shard_count, rs.ignore_msb};
+                            }
+                        };
```

Comment:
I'd like to extract this this `get_sharding_info` lambda to a function, but generally I think all the topology code deserves a larger refactor, at least moving it into its separate class(es) and all the lambdas to functions etc.. I'll probably do it at some point but in a separate PR.